[SPARK-12311][CORE] Restore previous value of "os.arch" property in test suites after forcing to set specific value to "os.arch" property

kiszk opened the pull request:
Restore the original value of os.arch property after each test

Since some of tests forced to set the specific value to os.arch property, we need to set the original value.

kiszk committed:
restore the original value of os.arch property after each test

Since some of tests forced to set the specific value to os.arch property, we need to set the original value.

AmplabJenkins commented:
Can one of the admins verify this patch?

JoshRosen commented:
I think that this was supposed to be handled by the `ResetSystemProperties` mixin; do either of these suites fail to call the superclass's `afterEach()` or `afterAll()` methods, by chance?

srowen commented:
Ah, I knew I was forgetting something. I agree this should already be handled. @kiszk are you sure it's actually not working or just thought it looked like it didn't?

holdenk commented:
@JoshRosen neither of those suites override the afterAll , but neither of them call the super's afterEach function (which they both override and that is where the ResetSystemProperties trait has it hook).

srowen commented:
@holdenk yeah I think that explains the problem in `BlockManagerSuite`, good catch. I'm not sure why the other one is a problem though; not overriding `afterEach` is fine as it means the superclass method does execute. Maybe that one isn't actually a problem @kiszk?

The general form of these overrides should be to run superclass init first, and superclass cleanup after:

```
override def beforeEach(): Unit = {
  super.beforeEach()
  ... init here ...
}

override def afterEach(): Unit = {
  try {
    ... cleanup here ...
  } finally {
    super.afterEach()
  }
}
```

Same for `beforeAll()` and `afterAll()`. _The `try-finally` isn't necessary if the cleanup really can't generate an exception._

We have a number of instances where this isn't done quite right across the code base,. AFAICT most don't matter. It's not that many though. @kiszk what do you think about correcting this pattern everywhere to head off potential bugs of this form later?

3ourroom commented:
NAVER - http://www.naver.com/
--------------------------------------------

3ourroom@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-12311][CORE] Restore previous value of "os.arch" property in test suites after forcing to set specific value to "os.arch" property (#10289)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

kiszk commented:
@srowen, OK, I will correct this pattern everywhere.

kiszk commented:
There are two potential bugs
1. A method of the super class is not called in `beforeEach()`, `afterEach()`, `beforeAll()`, and `afterAll()`
2. Although `System.setProperty()` is called, `ResetSystemProperties` is not mixed-in. 
I will correct them

kiszk committed:
restore the original values of System.property by using the ResetSystemProperties mixin

Call init/cleanup methods of super class

Use ResetSystemProperties if System.setProperty() is called

andrewor14 commented:
ok to test

SparkQA commented:
**[Test build #47884 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47884/consoleFull)** for PR 10289 at commit [`b6deb2f`](https://github.com/apache/spark/commit/b6deb2f59ad2245847cc15fefae8697649618cb9).

SparkQA commented:
**[Test build #47884 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47884/consoleFull)** for PR 10289 at commit [`b6deb2f`](https://github.com/apache/spark/commit/b6deb2f59ad2245847cc15fefae8697649618cb9).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class SecurityManagerSuite extends SparkFunSuite with ResetSystemProperties `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47884/
Test FAILed.

kiszk committed:
fix errors in lint-scala

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47893/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47894/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47895/
Test FAILed.

srowen commented on core/src/test/scala/org/apache/spark/util/collection/ExternalSorterSuite.scala:
```diff
@@ -235,7 +235,7 @@ class ExternalSorterSuite extends SparkFunSuite with LocalSparkContext {
     }
   }
-
+*/
```
Why is this big section commented out?

kiszk replied:
Sorry, this is my mistake. This file should not be modified.

srowen commented on mllib/src/test/scala/org/apache/spark/mllib/util/MLlibTestSparkContext.scala:
```diff
@@ -38,12 +38,15 @@ trait MLlibTestSparkContext extends BeforeAndAfterAll { self: Suite =>
+      }
     sc = null
-    super.afterAll()
+    } finally {
```
I think the line above needs an indent?

kiszk replied:
Yes, I will insert two spaces for an indent

srowen commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/BaseYarnClusterSuite.scala:
```diff
@@ -116,7 +120,7 @@ abstract class BaseYarnClusterSuite
   override def afterAll() {
     yarnCluster.stop()
-    System.clearProperty("SPARK_YARN_MODE")
+    System.setProperties(oldProperties)
```
This changes the logic of the test a bit; is this needed?

kiszk replied:
I am afraid that another property may be changes somewhere. I will use the original code since it works well now.

srowen replied:
It's probably a decent idea. I think you can leave the change.

srowen commented:
@kiszk I think it's looking good, minus some style problem that's making the build fail, and a few minor comments. I think it's pretty nice to clean up the original bug but also standardize and correct the behavior of these test lifecycle methods. It's worth a change that touches lots of files.

kiszk committed:
fix unexpect big section commented out

add spaces for the right indent

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47933/
Test FAILed.

SparkQA commented:
**[Test build #2224 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2224/consoleFull)** for PR 10289 at commit [`9b99df9`](https://github.com/apache/spark/commit/9b99df973f707e42e83926bfaf2f71a03538e65d).

SparkQA commented:
**[Test build #2224 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2224/consoleFull)** for PR 10289 at commit [`9b99df9`](https://github.com/apache/spark/commit/9b99df973f707e42e83926bfaf2f71a03538e65d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class SecurityManagerSuite extends SparkFunSuite with ResetSystemProperties `\n

SparkQA commented:
**[Test build #47995 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47995/consoleFull)** for PR 10289 at commit [`49b15f8`](https://github.com/apache/spark/commit/49b15f8745eee783b6fe6cafc476e057466c9962).

SparkQA commented:
**[Test build #47995 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47995/consoleFull)** for PR 10289 at commit [`49b15f8`](https://github.com/apache/spark/commit/49b15f8745eee783b6fe6cafc476e057466c9962).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class SecurityManagerSuite extends SparkFunSuite with ResetSystemProperties `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47995/
Test PASSed.

srowen commented:
@kiszk your last merge to this branch seems to have gone wrong. This has a lot of unrelated changes now. Can you try backing out the bad commits and force pushing? or getting the right commits to a new PR?

SparkQA commented:
**[Test build #48047 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48047/consoleFull)** for PR 10289 at commit [`7be29b1`](https://github.com/apache/spark/commit/7be29b179dc5a518e48372658ee965d3292b3021).

SparkQA commented:
**[Test build #48047 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48047/consoleFull)** for PR 10289 at commit [`7be29b1`](https://github.com/apache/spark/commit/7be29b179dc5a518e48372658ee965d3292b3021).
 * This patch **fails R style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:\n  * `class SecurityManagerSuite extends SparkFunSuite with ResetSystemProperties `\n  * `class CheckpointInputDStream(ssc_ : StreamingContext) extends InputDStream[Int](ssc_) `\n  * `  class FileInputDStreamCheckpointData extends DStreamCheckpointData(this) `\n

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48047/
Test FAILed.

kiszk commented:
I will fix this merge issue this weekend.

SparkQA commented:
**[Test build #48072 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48072/consoleFull)** for PR 10289 at commit [`43d1854`](https://github.com/apache/spark/commit/43d185493c8639a39dbdd54b54f8fe1285a89364).

SparkQA commented:
**[Test build #48072 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48072/consoleFull)** for PR 10289 at commit [`43d1854`](https://github.com/apache/spark/commit/43d185493c8639a39dbdd54b54f8fe1285a89364).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48072/
Test PASSed.

srowen commented:
OK, this looks good to me. It's fixing the original problem and fixing potential problems of the same form elsewhere.

andrewor14 commented:
Yeah, this looks fine. I wonder if there's something safer that we can do in the future. What I want is the following but I don't think scalastyle is powerful enough to support it:
```
// Allow
override def beforeAll(): Unit = {
  super.beforeAll()
  ...
}

// Disallow
override def beforeAll(): Unit = {
  // no super.beforeAll() here
  ...
}
```

srowen commented:
I suppose the superclass could catch the case where beforeAll wasn't called but afterAll was, but not the latter. I also don't know how to implement this without significant cleverness, which doesn't seem worth it.

andrewor14 commented:
Hm, I suppose we could do something in `SparkFunSuite`:

```
final def beforeAll(): Unit = {
  super.beforeAll()
  doBeforeAll()
}

protected override def doBeforeAll(): Unit
```
and then we'll use scalastyle to ban direct usages of `BeforeAndAfterAll`. Would that work? (I'm suggesting for a future patch)

andrewor14 commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/BaseYarnClusterSuite.scala:
```diff
@@ -116,7 +120,7 @@ abstract class BaseYarnClusterSuite
     yarnCluster.stop()
-    System.clearProperty("SPARK_YARN_MODE")
+    System.setProperties(oldProperties)
     super.afterAll()
```
this one needs to be in `try finally`

kiszk replied:
I agree. Since the original code did not use `try finally`, I did not use it, too. Should I use `try finally`?

srowen replied:
Yes, your change is adding this call though right? Ok to do it here too

kiszk replied:
I see, I will use `try finally` here

andrewor14 commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala:
```diff
@@ -43,12 +45,21 @@ import org.apache.spark.util.Utils
 
 class ClientSuite extends SparkFunSuite with Matchers with BeforeAndAfterAll {
 
+  var oldProperties: Properties = null
```
why not use `ResetSystemProperties` here?

kiszk replied:
This is because System.setProperty() is not used in this file.

srowen replied:
That's a fair point -- yes it's called two lines below, right? I think you could use the standard mechanism in the 2 places you're manually copying the sys properties?

kiszk replied:
Sorry, I made a mistake when I check all of the changes in this file. I will use `ResetSystemProperties` here.

srowen commented:
Yeah, though then you can't fix this for test classes that extend test classes... not sure if that exists today but could. My gut is that it's not worth it but yes that could help too.

kiszk committed:
update files based on code review

use try-finally as used in others

use ResetSystemProperties since System.setProperties() is called

SparkQA commented:
**[Test build #48202 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48202/consoleFull)** for PR 10289 at commit [`cb6a862`](https://github.com/apache/spark/commit/cb6a8626e9c29d8b4a1f4bd2f0163784cf306bb5).

SparkQA commented:
**[Test build #48202 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48202/consoleFull)** for PR 10289 at commit [`cb6a862`](https://github.com/apache/spark/commit/cb6a8626e9c29d8b4a1f4bd2f0163784cf306bb5).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class SecurityManagerSuite extends SparkFunSuite with ResetSystemProperties `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48202/
Test FAILed.

kiszk commented:
Can I retest this?  Timeout may occur in pyspark.

SparkQA commented:
**[Test build #2251 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2251/consoleFull)** for PR 10289 at commit [`cb6a862`](https://github.com/apache/spark/commit/cb6a8626e9c29d8b4a1f4bd2f0163784cf306bb5).

SparkQA commented:
**[Test build #2251 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2251/consoleFull)** for PR 10289 at commit [`cb6a862`](https://github.com/apache/spark/commit/cb6a8626e9c29d8b4a1f4bd2f0163784cf306bb5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class SecurityManagerSuite extends SparkFunSuite with ResetSystemProperties `\n

srowen commented:
I'm going to go ahead and merge this today

srowen commented:
Merged to master
